rfcs/0136-stabilize-incrementally.md

```diff
@@ -0,0 +1,387 @@
+The earlier steps have been carefully designed to be either easier or already started ("de-risked") to do our best to convince those that are concerned about the current "limbo" that stabilization will be gotten to no matter what, even if the splitting fails.
+
+## Difficulties in the roll-out
+
+Process concerns are addressed by having a clear process, with clear outcomes, before any stabilization is begun.
+That is the purpose of this RFC!
+
+## Flakes are criticized for eating other Nix features
+
+We do *not* propose adding delay to once-again separate those features from Flakes, as that would introduce more feature work which would delay stabilization and be unacceptable to the pro-Flakes faction.
+
+That said, an ancillary incremental beneift of incremental stabilization is to bolster a *sense* of layering in Nix's user interface that has been, according to some people, lost.
+Ensuring that these two subsets of the new CLI --- without Flakes, and without Nix language support --- do in fact make sense in isolation will provide a "scaffolding" upon which interested parties can later introduce generalized features like search and pure eval without Flakes.
+
+The hope is that such scaffolding will assuage this faction their concerns are heard without holding things up.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The main downside is a delay from the splitting Nix process, and then a delay between the stabilization steps.
+
+The fact that [RFC 134] is already mostly implemented, and in code review, is hopefully reason enough to believe it shouldn't take much longer.
+That maximum delay waiting for that to complete should be dwarfed by duration of time we've spent "in limbo" without a clear plan to move forward.
+We therefore think that is a small and reasonable price to pay for the benefit of community harmony.
+
+An addition, in the detailed design there is an escape hatch saying that blocking on the implementation of [RFC 134] can be reconsidered things do indeed take longer than foreseen.
+
+The delay of "auditing and refining" and shouldn't represent time "idle" from a stabilization perspective.
+As long as we are making progress stabilizing features and having healthy discussions, we don't see any problem.
+
+**Step 0** is also designed to take the pressure off these possible sources of delay, giving us *something* to work on that is not blocked on [RFC 134] or anything else.
+
+# Alternatives
+[alternatives]: #alternatives
+
+We could, of course, just "rip off the band-aid" and stabilize everything at once.
+The argument for that would be that enough time has passed and the concerns of (less numerous) long-time users are not important.
+But we think the plan here has little downsides; we can instead make everyone happy with only some delay.
+If that is true, why not do that instead!
+
+# Unresolved questions
```

Comment:
Let me preface this with appreciation for the work the author/shepherd team did to get things written down. The following is not to dismiss any of that, to the contrary, and is aiming to follow the practice of RFCs piling up considerations in order to help converging on a reasonable solution design.

Overall, I think the scope is too broad. There are many generally agreeable statements in there, but they tend to be vague and largely inconsequential. Regardless, I expect some of them to produce both some enlightening and some needless contention when scrutinized. I think this hasn't really happened yet, because the text is also still quite long for what it's trying to say, although it seems to have improved since the first versions.

Going in some more detail, the broad technical strokes seem fine, but could be a lot more concrete – if there is even anything to debate about them, that would be uncovered by delving into the actual design questions. That of course begs the question if there should be more RFCs detailing individual aspects (which is what Graham suggested in the post referenced here) or if some of those issues can be addressed as PRs. Many proposed statements can be managed easily within the Nix maintainers team and merged into official documentation directly.

Which brings me to the conceptual part: I understand this document is mainly about building trust and community consensus, but the problem statement and motivation is formulated only very weakly. What exactly is wrong with flakes not being stable? What exactly are the problems under discussion, and which trade-offs have to be considered for making a decision? Who are these "some people", and why does their opinion matter? To be precise, why is it that we can't just either remove the experimental flags xor remove the experimental feature itself? This would be important to state so readers can judge whether the problem is relevant and stated truthfully, whether the weighing of considerations is appropriate, and why the given conclusions are more reasonable than the alternatives.

Finally, I'm missing a consideration of project objectives and resource constraints. For instance, we can't really talk about timing. So far we didn't even manage to talk about order of work done, although we did succeed at agreeing on some priorities, such as improving testing and fixing bugs. Also, who is supposed to implement the changes? And wouldn't the proposed steps be taken regardless of this discussion anyway? Do Nix maintainers really need to request for comments to go about changes incrementally, as they do already? And for far-reaching changes, where I think it would be sensible to open a broader discussion, where is the threshold and should we take on even more of those instead of finishing one thing first?
For example, how do the issues in this RFC relate to substantial proposed changes such as layered stores or store ACLs?

Not addressing that would make this proposal, especially in light of its emphasis on popularity and people waiting for things, amount to saying "done when it's done, maybe never". Which would be absolutely no improvement over the current state of affairs.

Reply:
I've tried to formulate multiple answers to this comment because I felt it was quite a harsh take, but every time I tried to argue against it, I have to concede that you are correct.

This RFC has already influenced discussions AFAICT, but the substance is so broad that it's questionable whether merging it would have any impact internally beyond that. It's still valuable as a document to show what the current challenges are, but merging it seems to be inconsequential.

Reply:
I think the basic issue is that when I started writing this, I felt like very much an outsider to Nix's development, and the ideas expressed in here felt like outsider/underdog ideas. Now, I am on the Nix Team (!), and there are some other like-minded people on the Nix team, and the most pro-Flakes people don't seem to be pouncing on this RFC keeping high pressure to compromise.

If people don't show up to argue back, then yes it is tempting to revise this to be something a lot stiffer. E.g. my personal  "dream" roadmap:

 - Flakes, most of `nix-env`, `nix profile` and other ancillary features would be ripped out as separate downstream projects.
   - "Most of `nix-env`" would be legacy compat shim, more or less frozen interfaces
   - New tools (Flakes, `nix profile`, etc.) can do whatever they want.

 - Nix is stripped down to its core, focus is equally on Bindings, nice RPC protocols, SDKs, and CLI. The (freed from compat constraint) CLI would aim for being auto-generated (as @fricklerhandwerk mentions) over being ergonomic, whenever the two conflict
   - Also make plans to get off C++ within 2 years

- Major focus of any new features is low latency rebuilds (avoiding mass rebuilds, more fine-grained derivations, inotify) and thus pitching Nix as not just the "post-package manager" but a credible Bazel competitor.

- Windows support on the same tier as macOS.

- Standardize the "store layer" (store objects, derivations, etc.) with the IETF.

I guess I *could* go write *that* RFC. And yes, it would be a lot sharper. But I assume the vast majority of people here would be like "wtf is that I'm so confused?". I am I mistaken?

Reply:
It's tough, but I see what you mean. This is a good start. Having it codified would be a good checkpoint. There's nothing more for me to add. I think there are comments of me left, but after the recent changes I don't have anything more to say.

I guess it's time for FCP?

Reply:
Even if this RFC isn't great, it's not bad either, it definitely doesn't make things worse. I'd like accept this RFC as is just so we can move on and focus on more concrete efforts: https://github.com/NixOS/rfcs/pull/136#pullrequestreview-1530539951